---
title: "Comparison"
author: "Andreas Schneider"
date: "24/07/2019"
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
    theme: "spacelab"
    highlight: textmate
    df_print: paged
    code_folding: hide
    self_contained: false
    keep_md: false
    encoding: "UTF-8"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages}
library(dada2)
suppressPackageStartupMessages(library(phyloseq))
suppressPackageStartupMessages(library(Biostrings))
suppressPackageStartupMessages(library(VennDiagram))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(vegan))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(reshape2))
source("src/R/featureSelection.R")
```


# Comparison dada2 - vsearch

This document aims to show that dada2 ASV clustering (a more exact method than the traditional OTU clustering) provides the same results and conclusions as vsearch and OTU clustering used in Haas et al. 2018.

The soil samples are not included, since we don't have matching RNA-Seq data.

## Import and formatting of data

We start by reading in the preprocessed data. 

First 16S, needles and then roots. We remove the mock and negative control samples before we proceed.

```{r mock16Sneedle}
seq.tab_n.nochim <- readRDS("data/16S/Needles/dada2/seqtab_nochim.rds")
mock_n <- seq.tab_n.nochim[c(1,2,39),]
seq.tab_n.nochim2 <- seq.tab_n.nochim[-c(1,2,39),]
```

```{r mock16Sroot}
seq.tab_r.nochim <- readRDS("data/16S/Roots/dada2/seqtab_nochim.rds")
mock_r <- seq.tab_r.nochim[c(1,2,39),]
seq.tab_r.nochim2 <- seq.tab_r.nochim[-c(1,2,39),]
```

```{r prefiltering16S}
seq.tab_16s_all <- mergeSequenceTables(seq.tab_n.nochim2, seq.tab_r.nochim2)
##At this point we filter out ASVs with less than 10 from every sample and require a minimum abundance of 0.005% in "any of the three sample types".
seq.tab_16s_all[seq.tab_16s_all<10] <- 0

ccc_16S <- factor(paste(sapply(strsplit(rownames(seq.tab_16s_all), "[.]"), "[", c(1)), sapply(strsplit(rownames(seq.tab_16s_all), "[.]"), "[", c(2)), sep = "_"))
names(ccc_16S) <- rownames(seq.tab_16s_all)
ccc_16S <- ccc_16S[rownames(seq.tab_16s_all)]
seq.tab_16s_all2 <- seq.tab_16s_all[,featureSelectProp(as.matrix(t(seq.tab_16s_all)), ccc_16S, 0.00005)]

saveRDS(seq.tab_16s_all2, file = "data/16S/Both/dada2/seq.tab_all.rds")
dim(seq.tab_16s_all)
dim(seq.tab_16s_all2)
```

The low abundance filtering already decreases the number of ASVs from 17000 to 7000. After running taxonomy assignment, we can import the data into phyloseq and prune all Mitochondrial and chloroplast sequences.

```{r psimport16S}
ps_16s_all <- phyloseq(otu_table(readRDS("data/16S/Both/dada2/seq.tab_all.rds"), taxa_are_rows = FALSE),
                       sample_data(read.csv("doc/B16S_all.csv", row.names = 1)),
                       tax_table(readRDS("data/16S/Both/dada2/taxa.rds")))
```

```{r filtering16S}
dna_16s_all <- DNAStringSet(taxa_names(ps_16s_all))
names(dna_16s_all) <- taxa_names(ps_16s_all)
ps_16s_all <- merge_phyloseq(ps_16s_all, dna_16s_all)
taxa_names(ps_16s_all) <- paste0("ASV", seq(ntaxa(ps_16s_all)))
names(dna_16s_all) <- paste0("ASV", seq(ntaxa(ps_16s_all)))

taxa_16s_all <- as.data.frame(tax_table(ps_16s_all))
taxa_16s_all[is.na(taxa_16s_all)] <- "Unknown"
valid_asv_16s <- rownames(taxa_16s_all[taxa_16s_all$Family!="Mitochondria"&
                                taxa_16s_all$Order!="Chloroplast",])

ps_16s_all <- prune_taxa(valid_asv_16s, ps_16s_all)
dim(otu_table(ps_16s_all))
```

About 6000 ASVs survive the Chloro/Mito pruning.

Now we are ready to import the ITS data. We import both needles and root data, and after cleaning up the sequences (we select what has been cut out by ITSx, this should allow for more accurate merging of ASVs), we merge them and run the taxonomy assignment on the merged matrix.

```{r importITSneedle}
seq.tab_n_i.nochim <- readRDS("data/ITS/Needles/dada2/seqtab_nochim.rds")
seq.tab_n_i.nochim2 <- seq.tab_n_i.nochim[,!(colnames(seq.tab_n_i.nochim) %in% readDNAStringSet("data/ITS/Needles/ITSx/refseq-ITSx.fa_no_detections.fasta"))]

colnames(seq.tab_n_i.nochim2)[which(paste0("ASV", match(colnames(seq.tab_n_i.nochim2), colnames(seq.tab_n_i.nochim))) %in% gsub("(ASV\\d+)\\|.*", "\\1", names(readDNAStringSet("data/ITS/Needles/ITSx/refseq-ITSx.fa.ITS1.fasta"))))] <- readDNAStringSet("data/ITS/Needles/ITSx/refseq-ITSx.fa.ITS1.fasta")
mock_n_i <- seq.tab_n_i.nochim2["its.mock.none.t5.0",]
seq.tab_n_i.nochim2 <- seq.tab_n_i.nochim2[-1,]
```

```{r importITSroot}
seq.tab_r_i.nochim <- readRDS("data/ITS/Roots/dada2/seqtab_nochim.rds")
seq.tab_r_i.nochim2 <- seq.tab_r_i.nochim[,!(colnames(seq.tab_r_i.nochim) %in% readDNAStringSet("data/ITS/Roots/ITSx/refseq-ITSx.fa_no_detections.fasta"))]

colnames(seq.tab_r_i.nochim2)[which(paste0("ASV", match(colnames(seq.tab_r_i.nochim2), colnames(seq.tab_r_i.nochim))) %in% gsub("(ASV\\d+)\\|.*", "\\1", names(readDNAStringSet("data/ITS/Roots/ITSx/refseq-ITSx.fa.ITS1.fasta"))))] <- readDNAStringSet("data/ITS/Roots/ITSx/refseq-ITSx.fa.ITS1.fasta")
mock_r_i <- seq.tab_r_i.nochim2["its.mock.none.t5.0",]
seq.tab_r_i.nochim2 <- seq.tab_r_i.nochim2[-1,]
```

Now we can merge the roots and needles datasets before the taxonomy assignment.

```{r filteringITS}
seq.tab_its_all <- mergeSequenceTables(seq.tab_n_i.nochim2, seq.tab_r_i.nochim2)
###Now some of the sequences will be shorter than 50nts, so we remove those (necessary for tax assignment)
seq.tab_its_all <- seq.tab_its_all[,which(nchar(colnames(seq.tab_its_all))>50)]
##ALso at this point we filter out ASVs with less than 10 from every sample and require a minimum abundance of 0.005% in "any of the three sample types".
seq.tab_its_all[seq.tab_its_all<10] <- 0
ccc <- factor(paste(sapply(strsplit(rownames(seq.tab_its_all), "[.]"), "[", c(2)), sapply(strsplit(rownames(seq.tab_its_all), "[.]"), "[", c(3)), sep = "_"))
names(ccc) <- rownames(seq.tab_its_all)
ccc <- ccc[rownames(seq.tab_its_all)]
seq.tab_its_all2 <- seq.tab_its_all[,featureSelectProp(as.matrix(t(seq.tab_its_all)), ccc, 0.00005)]

saveRDS(seq.tab_its_all, file = "data/ITS/Both/dada2/seq.tab_all.rds")

```

We run the taxonomy assignment through slurm. At this point the data is ready to be imported into phyloseq.

```{r psimportITS}
ps_its_all <- phyloseq(otu_table(readRDS("data/ITS/Both/dada2/seq.tab_all.rds"), taxa_are_rows = FALSE),
                       sample_data(read.csv("doc/ITS_all.csv", row.names = 1)),
                       tax_table(readRDS("data/ITS/Both/dada2/taxa.rds")))

```

Now we change the ASV names to ASV1, ASV2, etc and save the sequences in a separate slot. This will also be useful to filter them with ITSx.

```{r inclDNAITS}
dna_its_all <- DNAStringSet(taxa_names(ps_its_all))
names(dna_its_all) <- taxa_names(ps_its_all)
ps_its_all <- merge_phyloseq(ps_its_all, dna_its_all)
taxa_names(ps_its_all) <- paste0("ASV", seq(ntaxa(ps_its_all)))
```

## Fig. 1

Original figure:

![](`r "https://ars.els-cdn.com/content/image/1-s2.0-S0038071718302335-gr1.jpg"`)

We start by creating Venns of the ASVs from roots and needles, corresponding to fig. 1a,b. For Fig. 1 we use only the control samples

```{r plotVenns, out.width='50%',fig.show='hold'}
#Create list of ASVs for roots and needles
#16S
asv_list_16s <- list()
#NO
asv_list_16s[[1]] <- colnames(otu_table(ps_16s_all))[colSums(otu_table(ps_16s_all)[25:36,])>0]
asv_list_16s[[2]] <- colnames(otu_table(ps_16s_all))[colSums(otu_table(ps_16s_all)[61:72])>0]
png_16s <- "results/plots/venn_16s.png"

#ITS
asv_list_its <- list()
asv_list_its[[1]] <- colnames(otu_table(ps_its_all))[colSums(otu_table(ps_its_all)[25:36,])>0]
asv_list_its[[2]] <- colnames(otu_table(ps_its_all))[colSums(otu_table(ps_its_all)[61:72])>0]
png_its <- "results/plots/venn_its.png"
#suppress log files
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
pal=brewer.pal(8,"Dark2")
par(mfrow=c(1,1))
par(mar=c(0.5,0.5,0.5,0.5))
png(png_16s)
grid.newpage()
grid.draw(venn.diagram(asv_list_its,
                       main = "Fungi",
                       main.cex = 3,
                       height = 8,
                       width = 8,
                       filename=NULL,
                       category.names = c("Needles","Roots"),
                       fill=pal[1:2]))
dev.off()

png(png_its)
grid.newpage()
grid.draw(venn.diagram(asv_list_16s,
                       main = "Bacteria",
                       main.cex = 3,
                       height = 8,
                       width = 8,
                       filename=NULL,
                       category.names = c("Needles","Roots"),
                       fill=pal[1:2]))
dev.off()
knitr::include_graphics(c(png_16s, png_its))
```

The proportions are very similar, but the absolute numbers are of course higher because the ASVs are basically clustered with 100% similarity (vs 97/95%).

We continue with the Shannon diversity boxplots of the control samples, for this have to rarefy the data.


```{r Shannon1}
ctrl_tru <- grepl("ctrl", rownames(otu_table(ps_its_all)))
ps_its_all_c <- prune_samples(ctrl_tru, ps_its_all)
mat_its_rar_c <- rrarefy(as.matrix(otu_table(ps_its_all_c)), min(rowSums(otu_table(ps_its_all_c))))
shannon1_its <- diversity(mat_its_rar_c)
shannon1_its <- melt(shannon1_its)
shannon1_its$Type <- factor(read.csv("doc/ITS_all.csv")[ctrl_tru,2], levels = c("root", "needle"))

shannon1_its$Timepoint <- as.factor(read.csv("doc/ITS_all.csv")[ctrl_tru,5])
shannon1_its$Organism <- as.factor("Fungi")

ctrl_tru2 <- grepl("ctrl", rownames(otu_table(ps_16s_all)))
ps_16s_all_c <- prune_samples(ctrl_tru2, ps_16s_all)
mat_16s_rar_c <- rrarefy(as.matrix(otu_table(ps_16s_all_c)), min(rowSums(otu_table(ps_16s_all_c))))
shannon1_16s <- diversity(mat_16s_rar_c)
shannon1_16s <- melt(shannon1_16s)
shannon1_16s$Type <- factor(read.csv("doc/B16S_all.csv")[ctrl_tru2,2], levels = c("root", "needle"))
shannon1_16s$Timepoint <- as.factor(read.csv("doc/B16S_all.csv")[ctrl_tru2,5])
shannon1_16s$Organism <- as.factor("Bacteria")

shannon1_all <- rbind(shannon1_its, shannon1_16s)
colnames(shannon1_all) <- c("Shannon", "Type", "Time", "Organism")
shannon1_all <- as.data.frame(shannon1_all)

ggformat <- theme_classic()+
  theme(axis.text = element_text(face = "bold", size = 15),
        axis.text.x = element_text(vjust = 1),
        axis.line = element_line(size = 1, linetype = "solid"),
        axis.title = element_text(size = 15, face = "bold"),
        axis.ticks = element_line(size = 1, linetype = "solid"),
        legend.text = element_text(size=15, face = "bold"),
        legend.title = element_text(size = 17, face = "bold"),
        plot.title = element_text(size = 21, face = "bold"),
        strip.text = element_text(size = 19, face = "bold"))

ggplot(shannon1_all, aes(x = Type, y = Shannon, fill = Time))+
  geom_boxplot()+
  facet_grid(~Organism)+
 ggformat
```

Looks very similar to the original figure (except for some minor differences between the timepoints).

Next we have to perform the same statistical tests to deter
