<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Andreas Schneider" />


<title>Comparison</title>

<script src="Comparison_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Comparison_files/bootstrap-3.3.5/css/spacelab.min.css" rel="stylesheet" />
<script src="Comparison_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Comparison_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Comparison_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="Comparison_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="Comparison_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="Comparison_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="Comparison_files/navigation-1.1/tabsets.js"></script>
<script src="Comparison_files/navigation-1.1/codefolding.js"></script>
<link href="Comparison_files/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="Comparison_files/highlightjs-9.12.0/highlight.js"></script>
<link href="Comparison_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="Comparison_files/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Comparison</h1>
<h4 class="author">Andreas Schneider</h4>
<h4 class="date">24/07/2019</h4>

</div>


<pre class="r"><code>library(dada2)</code></pre>
<pre><code>## Loading required package: Rcpp</code></pre>
<pre class="r"><code>suppressPackageStartupMessages(library(phyloseq))
suppressPackageStartupMessages(library(Biostrings))
suppressPackageStartupMessages(library(VennDiagram))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(vegan))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(reshape2))
source(&quot;src/R/featureSelection.R&quot;)</code></pre>
<div id="comparison-dada2---vsearch" class="section level1">
<h1><span class="header-section-number">1</span> Comparison dada2 - vsearch</h1>
<p>This document aims to show that dada2 ASV clustering (a more exact method than the traditional OTU clustering) provides the same results and conclusions as vsearch and OTU clustering used in Haas et al. 2018.</p>
<p>The soil samples are not included, since we don’t have matching RNA-Seq data.</p>
<div id="import-and-formatting-of-data" class="section level2">
<h2><span class="header-section-number">1.1</span> Import and formatting of data</h2>
<p>We start by reading in the preprocessed data.</p>
<p>First 16S, needles and then roots. We remove the mock and negative control samples before we proceed.</p>
<pre class="r"><code>seq.tab_n.nochim &lt;- readRDS(&quot;data/16S/Needles/dada2/seqtab_nochim.rds&quot;)
mock_n &lt;- seq.tab_n.nochim[c(1,2,39),]
seq.tab_n.nochim2 &lt;- seq.tab_n.nochim[-c(1,2,39),]</code></pre>
<pre class="r"><code>seq.tab_r.nochim &lt;- readRDS(&quot;data/16S/Roots/dada2/seqtab_nochim.rds&quot;)
mock_r &lt;- seq.tab_r.nochim[c(1,2,39),]
seq.tab_r.nochim2 &lt;- seq.tab_r.nochim[-c(1,2,39),]</code></pre>
<pre class="r"><code>seq.tab_16s_all &lt;- mergeSequenceTables(seq.tab_n.nochim2, seq.tab_r.nochim2)
##At this point we filter out ASVs with less than 10 from every sample and require a minimum abundance of 0.005% in &quot;any of the three sample types&quot;.
seq.tab_16s_all[seq.tab_16s_all&lt;10] &lt;- 0

ccc_16S &lt;- factor(paste(sapply(strsplit(rownames(seq.tab_16s_all), &quot;[.]&quot;), &quot;[&quot;, c(1)), sapply(strsplit(rownames(seq.tab_16s_all), &quot;[.]&quot;), &quot;[&quot;, c(2)), sep = &quot;_&quot;))
names(ccc_16S) &lt;- rownames(seq.tab_16s_all)
ccc_16S &lt;- ccc_16S[rownames(seq.tab_16s_all)]
seq.tab_16s_all2 &lt;- seq.tab_16s_all[,featureSelectProp(as.matrix(t(seq.tab_16s_all)), ccc_16S, 0.00005)]

saveRDS(seq.tab_16s_all2, file = &quot;data/16S/Both/dada2/seq.tab_all.rds&quot;)
dim(seq.tab_16s_all)</code></pre>
<pre><code>## [1]    72 17550</code></pre>
<pre class="r"><code>dim(seq.tab_16s_all2)</code></pre>
<pre><code>## [1]   72 7615</code></pre>
<p>The low abundance filtering already decreases the number of ASVs from 17000 to 7000. After running taxonomy assignment, we can import the data into phyloseq and prune all Mitochondrial and chloroplast sequences.</p>
<pre class="r"><code>ps_16s_all &lt;- phyloseq(otu_table(readRDS(&quot;data/16S/Both/dada2/seq.tab_all.rds&quot;), taxa_are_rows = FALSE),
                       sample_data(read.csv(&quot;doc/B16S_all.csv&quot;, row.names = 1)),
                       tax_table(readRDS(&quot;data/16S/Both/dada2/taxa.rds&quot;)))</code></pre>
<pre class="r"><code>dna_16s_all &lt;- DNAStringSet(taxa_names(ps_16s_all))
names(dna_16s_all) &lt;- taxa_names(ps_16s_all)
ps_16s_all &lt;- merge_phyloseq(ps_16s_all, dna_16s_all)
taxa_names(ps_16s_all) &lt;- paste0(&quot;ASV&quot;, seq(ntaxa(ps_16s_all)))
names(dna_16s_all) &lt;- paste0(&quot;ASV&quot;, seq(ntaxa(ps_16s_all)))

taxa_16s_all &lt;- as.data.frame(tax_table(ps_16s_all))
taxa_16s_all[is.na(taxa_16s_all)] &lt;- &quot;Unknown&quot;
valid_asv_16s &lt;- rownames(taxa_16s_all[taxa_16s_all$Family!=&quot;Mitochondria&quot;&amp;
                                taxa_16s_all$Order!=&quot;Chloroplast&quot;,])

ps_16s_all &lt;- prune_taxa(valid_asv_16s, ps_16s_all)
dim(otu_table(ps_16s_all))</code></pre>
<pre><code>## [1]   72 6111</code></pre>
<p>About 6000 ASVs survive the Chloro/Mito pruning.</p>
<p>Now we are ready to import the ITS data. We import both needles and root data, and after cleaning up the sequences (we select what has been cut out by ITSx, this should allow for more accurate merging of ASVs), we merge them and run the taxonomy assignment on the merged matrix.</p>
<pre class="r"><code>seq.tab_n_i.nochim &lt;- readRDS(&quot;data/ITS/Needles/dada2/seqtab_nochim.rds&quot;)
seq.tab_n_i.nochim2 &lt;- seq.tab_n_i.nochim[,!(colnames(seq.tab_n_i.nochim) %in% readDNAStringSet(&quot;data/ITS/Needles/ITSx/refseq-ITSx.fa_no_detections.fasta&quot;))]

colnames(seq.tab_n_i.nochim2)[which(paste0(&quot;ASV&quot;, match(colnames(seq.tab_n_i.nochim2), colnames(seq.tab_n_i.nochim))) %in% gsub(&quot;(ASV\\d+)\\|.*&quot;, &quot;\\1&quot;, names(readDNAStringSet(&quot;data/ITS/Needles/ITSx/refseq-ITSx.fa.ITS1.fasta&quot;))))] &lt;- readDNAStringSet(&quot;data/ITS/Needles/ITSx/refseq-ITSx.fa.ITS1.fasta&quot;)
mock_n_i &lt;- seq.tab_n_i.nochim2[&quot;its.mock.none.t5.0&quot;,]
seq.tab_n_i.nochim2 &lt;- seq.tab_n_i.nochim2[-1,]</code></pre>
<pre class="r"><code>seq.tab_r_i.nochim &lt;- readRDS(&quot;data/ITS/Roots/dada2/seqtab_nochim.rds&quot;)
seq.tab_r_i.nochim2 &lt;- seq.tab_r_i.nochim[,!(colnames(seq.tab_r_i.nochim) %in% readDNAStringSet(&quot;data/ITS/Roots/ITSx/refseq-ITSx.fa_no_detections.fasta&quot;))]

colnames(seq.tab_r_i.nochim2)[which(paste0(&quot;ASV&quot;, match(colnames(seq.tab_r_i.nochim2), colnames(seq.tab_r_i.nochim))) %in% gsub(&quot;(ASV\\d+)\\|.*&quot;, &quot;\\1&quot;, names(readDNAStringSet(&quot;data/ITS/Roots/ITSx/refseq-ITSx.fa.ITS1.fasta&quot;))))] &lt;- readDNAStringSet(&quot;data/ITS/Roots/ITSx/refseq-ITSx.fa.ITS1.fasta&quot;)
mock_r_i &lt;- seq.tab_r_i.nochim2[&quot;its.mock.none.t5.0&quot;,]
seq.tab_r_i.nochim2 &lt;- seq.tab_r_i.nochim2[-1,]</code></pre>
<p>Now we can merge the roots and needles datasets before the taxonomy assignment.</p>
<pre class="r"><code>seq.tab_its_all &lt;- mergeSequenceTables(seq.tab_n_i.nochim2, seq.tab_r_i.nochim2)
###Now some of the sequences will be shorter than 50nts, so we remove those (necessary for tax assignment)
seq.tab_its_all &lt;- seq.tab_its_all[,which(nchar(colnames(seq.tab_its_all))&gt;50)]
##ALso at this point we filter out ASVs with less than 10 from every sample and require a minimum abundance of 0.005% in &quot;any of the three sample types&quot;.
seq.tab_its_all[seq.tab_its_all&lt;10] &lt;- 0
ccc &lt;- factor(paste(sapply(strsplit(rownames(seq.tab_its_all), &quot;[.]&quot;), &quot;[&quot;, c(2)), sapply(strsplit(rownames(seq.tab_its_all), &quot;[.]&quot;), &quot;[&quot;, c(3)), sep = &quot;_&quot;))
names(ccc) &lt;- rownames(seq.tab_its_all)
ccc &lt;- ccc[rownames(seq.tab_its_all)]
seq.tab_its_all2 &lt;- seq.tab_its_all[,featureSelectProp(as.matrix(t(seq.tab_its_all)), ccc, 0.00005)]

saveRDS(seq.tab_its_all, file = &quot;data/ITS/Both/dada2/seq.tab_all.rds&quot;)</code></pre>
<p>We run the taxonomy assignment through slurm. At this point the data is ready to be imported into phyloseq.</p>
<pre class="r"><code>ps_its_all &lt;- phyloseq(otu_table(readRDS(&quot;data/ITS/Both/dada2/seq.tab_all.rds&quot;), taxa_are_rows = FALSE),
                       sample_data(read.csv(&quot;doc/ITS_all.csv&quot;, row.names = 1)),
                       tax_table(readRDS(&quot;data/ITS/Both/dada2/taxa.rds&quot;)))</code></pre>
<p>Now we change the ASV names to ASV1, ASV2, etc and save the sequences in a separate slot. This will also be useful to filter them with ITSx.</p>
<pre class="r"><code>dna_its_all &lt;- DNAStringSet(taxa_names(ps_its_all))
names(dna_its_all) &lt;- taxa_names(ps_its_all)
ps_its_all &lt;- merge_phyloseq(ps_its_all, dna_its_all)
taxa_names(ps_its_all) &lt;- paste0(&quot;ASV&quot;, seq(ntaxa(ps_its_all)))</code></pre>
</div>
<div id="fig.-1" class="section level2">
<h2><span class="header-section-number">1.2</span> Fig. 1</h2>
<p>Original figure:</p>
<p><img src="https://ars.els-cdn.com/content/image/1-s2.0-S0038071718302335-gr1.jpg" /></p>
<p>We start by creating Venns of the ASVs from roots and needles, corresponding to fig. 1a,b. For Fig. 1 we use only the control samples</p>
<pre class="r"><code>#Create list of ASVs for roots and needles
#16S
asv_list_16s &lt;- list()
#NO
asv_list_16s[[1]] &lt;- colnames(otu_table(ps_16s_all))[colSums(otu_table(ps_16s_all)[25:36,])&gt;0]
asv_list_16s[[2]] &lt;- colnames(otu_table(ps_16s_all))[colSums(otu_table(ps_16s_all)[61:72])&gt;0]
png_16s &lt;- &quot;results/plots/venn_16s.png&quot;

#ITS
asv_list_its &lt;- list()
asv_list_its[[1]] &lt;- colnames(otu_table(ps_its_all))[colSums(otu_table(ps_its_all)[25:36,])&gt;0]
asv_list_its[[2]] &lt;- colnames(otu_table(ps_its_all))[colSums(otu_table(ps_its_all)[61:72])&gt;0]
png_its &lt;- &quot;results/plots/venn_its.png&quot;
#suppress log files
futile.logger::flog.threshold(futile.logger::ERROR, name = &quot;VennDiagramLogger&quot;)</code></pre>
<pre><code>## NULL</code></pre>
<pre class="r"><code>pal=brewer.pal(8,&quot;Dark2&quot;)
par(mfrow=c(1,1))
par(mar=c(0.5,0.5,0.5,0.5))
png(png_16s)
grid.newpage()
grid.draw(venn.diagram(asv_list_its,
                       main = &quot;Fungi&quot;,
                       main.cex = 3,
                       height = 8,
                       width = 8,
                       filename=NULL,
                       category.names = c(&quot;Needles&quot;,&quot;Roots&quot;),
                       fill=pal[1:2]))
dev.off()</code></pre>
<pre><code>## png 
##   2</code></pre>
<pre class="r"><code>png(png_its)
grid.newpage()
grid.draw(venn.diagram(asv_list_16s,
                       main = &quot;Bacteria&quot;,
                       main.cex = 3,
                       height = 8,
                       width = 8,
                       filename=NULL,
                       category.names = c(&quot;Needles&quot;,&quot;Roots&quot;),
                       fill=pal[1:2]))
dev.off()</code></pre>
<pre><code>## png 
##   2</code></pre>
<pre class="r"><code>knitr::include_graphics(c(png_16s, png_its))</code></pre>
<p><img src="results/plots/venn_16s.png" width="50%" /><img src="results/plots/venn_its.png" width="50%" /></p>
<p>The proportions are very similar, but the absolute numbers are of course higher because the ASVs are basically clustered with 100% similarity (vs 97/95%).</p>
<p>We continue with the Shannon diversity boxplots of the control samples, for this have to rarefy the data.</p>
<pre class="r"><code>ctrl_tru &lt;- grepl(&quot;ctrl&quot;, rownames(otu_table(ps_its_all)))
ps_its_all_c &lt;- prune_samples(ctrl_tru, ps_its_all)
mat_its_rar_c &lt;- rrarefy(as.matrix(otu_table(ps_its_all_c)), min(rowSums(otu_table(ps_its_all_c))))
shannon1_its &lt;- diversity(mat_its_rar_c)
shannon1_its &lt;- melt(shannon1_its)
shannon1_its$Type &lt;- factor(read.csv(&quot;doc/ITS_all.csv&quot;)[ctrl_tru,2], levels = c(&quot;root&quot;, &quot;needle&quot;))

shannon1_its$Timepoint &lt;- as.factor(read.csv(&quot;doc/ITS_all.csv&quot;)[ctrl_tru,5])
shannon1_its$Organism &lt;- as.factor(&quot;Fungi&quot;)

ctrl_tru2 &lt;- grepl(&quot;ctrl&quot;, rownames(otu_table(ps_16s_all)))
ps_16s_all_c &lt;- prune_samples(ctrl_tru2, ps_16s_all)
mat_16s_rar_c &lt;- rrarefy(as.matrix(otu_table(ps_16s_all_c)), min(rowSums(otu_table(ps_16s_all_c))))
shannon1_16s &lt;- diversity(mat_16s_rar_c)
shannon1_16s &lt;- melt(shannon1_16s)
shannon1_16s$Type &lt;- factor(read.csv(&quot;doc/B16S_all.csv&quot;)[ctrl_tru2,2], levels = c(&quot;root&quot;, &quot;needle&quot;))
shannon1_16s$Timepoint &lt;- as.factor(read.csv(&quot;doc/B16S_all.csv&quot;)[ctrl_tru2,5])
shannon1_16s$Organism &lt;- as.factor(&quot;Bacteria&quot;)

shannon1_all &lt;- rbind(shannon1_its, shannon1_16s)
colnames(shannon1_all) &lt;- c(&quot;Shannon&quot;, &quot;Type&quot;, &quot;Time&quot;, &quot;Organism&quot;)
shannon1_all &lt;- as.data.frame(shannon1_all)

ggformat &lt;- theme_classic()+
  theme(axis.text = element_text(face = &quot;bold&quot;, size = 15),
        axis.text.x = element_text(vjust = 1),
        axis.line = element_line(size = 1, linetype = &quot;solid&quot;),
        axis.title = element_text(size = 15, face = &quot;bold&quot;),
        axis.ticks = element_line(size = 1, linetype = &quot;solid&quot;),
        legend.text = element_text(size=15, face = &quot;bold&quot;),
        legend.title = element_text(size = 17, face = &quot;bold&quot;),
        plot.title = element_text(size = 21, face = &quot;bold&quot;),
        strip.text = element_text(size = 19, face = &quot;bold&quot;))

ggplot(shannon1_all, aes(x = Type, y = Shannon, fill = Time))+
  geom_boxplot()+
  facet_grid(~Organism)+
 ggformat</code></pre>
<p><img src="Comparison_files/figure-html/Shannon1-1.png" width="672" /></p>
<p>Looks very similar to the original figure (except for some minor differences between the timepoints).</p>
<p>Next we have to perform the same statistical tests to deter</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
