---
title: "dada2 preprocessing of 2012 data"
author: "Andreas Schneider"
date: "18/07/2019"
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
    theme: "spacelab"
    highlight: textmate
    df_print: paged
    code_folding: hide
    self_contained: false
    keep_md: false
    encoding: "UTF-8"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r import}
library(here)
library(dada2)
```

# Introduction

The data will be imported into R and clustered to ASV (amplicon sequencing variants) with the help of the dada2 R package

# Import of Data

Here the data is imported into R, and a quality check performed on some of the samples.

## 16S

### Needles
```{r cache=TRUE}
path_n <- here("data/16S/Needles/DeML_pooled/")

fnFs_n <- sort(list.files(path_n, pattern = "R1.fastq.gz", full.names = TRUE))
fnRs_n <- sort(list.files(path_n, pattern = "R2.fastq.gz", full.names = TRUE))

sample.names_n <- sapply(strsplit(basename(fnFs_n), "_"), `[`, 1)

plotQualityProfile(fnFs_n[1:4])
plotQualityProfile(fnRs_n[1:4])
```

The quality scores drop below 30 on average around 150, and we need more than 150bp per read for them to overlap properly. Thus we choose a cutoff at 190 for both.

In the next step we create a folder and filenames for the filtered files, and perform the actual filtering. Since this step is relatively fast we can run it inside the Rmarkdown document.

```{r cache=TRUE}
filtFs_n <- file.path(path_n, "filtered_F", paste0(sample.names_n, "_F_filt.fastq.gz"))
filtRs_n <- file.path(path_n, "filtered_R", paste0(sample.names_n, "_R_filt.fastq.gz"))
names(filtFs_n) <- sample.names_n
names(filtRs_n) <- sample.names_n

out <- filterAndTrim(fnFs_n, filtFs_n, fnRs_n, filtRs_n, truncLen=c(190,190),
                     maxN=0, maxEE=c(3,3), truncQ=2, rm.phix=TRUE,
                     compress=TRUE, multithread=TRUE)
head(out, n = 10)

```

```{bash}
module load bioinfo-tools R
#Rscript Error_Learning.R 
```


### Roots
```{r}
path_r <- here("data/16S/Roots/DeML_pooled/")
```


## ITS

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
